### ISSUE TITLE
Display tablatures

### ISSUE THREAD

**Issue by pikkle** [opensheetmusicdisplay/opensheetmusicdisplay#126]
VexFlow allows the displaying of guitar tablatures, and MusicXML adds a specific notation to a note for this ([documentation here](http://www.musicxml.com/tutorial/tablature/fret-and-string/)). Is there any plan to add either a single tablature display or a tablature-musicsheet combo (as displayed on [VexFlow main page](http://www.vexflow.com/)) ?

**Comment by bohoffi** [comment]
Like the main page is proving VexFlow already includes such a functionality.
The easiest way would be to call
`Vex.Flow.Formatter.FormatAndDrawTab(context, tabStave, stave, tabNotes, staveNotes, true, { align_rests: true });`
but this isn't a suitable be the solution for this issue.
The code inside the stated function has to be adapted.

If I may I would like to get this working in a fork.

**Comment by bohoffi** [comment]
**UPDATE:**
![2017-05-17_13h15_56](https://cloud.githubusercontent.com/assets/9944566/26151408/0dfe42b2-3b03-11e7-9775-aeffc9883331.png)

![2017-05-17_13h07_27](https://cloud.githubusercontent.com/assets/9944566/26151170/d3b9b664-3b01-11e7-95c9-5831165b2f6c.png)

Missing technical notation will be filtered (or better transformed to TabNote (`{ fret: "", str: "" }`). This is done to supress TAB rendering for files without any note with technical notation.
Currently this leads to the rendered results above:
- Sonatina
    - there are no notes with technical notation in the XML file
        - no TAB measures are rendered
- Saltarello
    - if there is at least 1 note with a technical notation available
        - the measure will be rendered
        - notes with missing technical notation will be rendered as `blanks` to "fill" gaps
        - there are still issues with the positioning even with only 1 voice (you could imagine what it will look like with multiple voices (e.g. piano))

It will be a pleasure to get some help with the positioning.

The current state was pushed to my fork.

**Comment by sebastianhaas** [comment]
Thanks @pikkle for bringing this up, and obviously @bohoffi for providing such a great starting point! I would very much like to see tabs support for OSMD, so I hope we can get your work there. 

Right now, I'm lacking basic understanding of how tabs work in both MusicXML and vexflow, but I'll dig into it next week. Would you be able to outline the main tasks/issues that need to be resolved?

Thank you very much!

**Comment by bohoffi** [comment]
Hi @sebastianhaas ,

of course I can:
- parse the `technical` tag ([doc](https://usermanuals.musicxml.com/MusicXML/Content/EL-MusicXML-technical.htm)) which **could** be inside `<note><notations></notations></note>`
    - **could** because if I understand the documentation correctly the tag is not required
    - **could** because there may be no `technical` tag on instruments without strings (piano, trumpet, etc.)
    - **could** because the `technical`tag may miss even on stringed instruments
        - e.g. no value while creating the MusicXML file
- parse the `string` and `fret` tags inside the `technical` tag
    - **but** I don't no what will be the case on fretless instruments (e.g. fretless bass) => `fret` will be missing?
    - **but** these two tags aren't required too
- if the string and the fret were successfully parsed, a Vex.Flow.TabNote can be created
- a `VexFlowMeasure` has to have:
    - a second stave => `TabStave`
    - a second property to hold the voices for the tab notes
    - draw the "TabVoices" to the `TabStave` (`this.vfTabVoices[voiceID].draw(ctx, this.tabStave);`)


If I left anything unclear, please let me know.


**Bonus thoughts**
Let's imagine we have a MusicXML file and the instruments are missing the above mentioned tags (`technical`, `string`, `fret`). We could be able to render the TabNotes even without these tags ;)
It such a case we can - if available - utilize the [String Tuning](http://www.musicxml.com/tutorial/tablature/string-tuning/). A quick idea to get the string and the fret from the tuning:
- provided tuning: `['E4', 'B3', 'G3', 'D3', 'A2', 'E2']` (Standard 6-string guitar tuning (highest to lowest string))
- generate a scale array: `['A0', 'A#0', 'B0', ..., 'B7', 'C8']` (standard 88-key piano scale)
- get **string** numbers (`index + 1`)
- get the scale indices of the string base notes (e.g. `'E4' => 43`)
- get the scale index of the note to play (e.g. `'F3' => 32`)
- filter all "string" which base note's scale index is <= the note's scale index (result: `['D3', 'A2', 'E2']`)
- **optional** limit the filtered string using a `maxFretCount` (`scale.slice(baseNoteScaleIndex).indexOf(note) <= maxFretCount`)
- get **fret** (`scale.slice(fb.baseNoteScaleIndex).indexOf(note)`)

Using the described way you could receive multiple `{ string: number; fret: number; }` objects. (As a guitarrist I would prefer the one with the lowest fret ;) )
If you are interested I could provide a gist showing the code I use for this "calculation".

**Comment by bohoffi** [comment]
Ok, I couldn't resist:
[getStringAndFretFromKeyAndTuning.ts](https://gist.github.com/bohoffi/637f3261d7f7d3a65fda8ee15b24f97a)

**Comment by matt-uib** [comment]
@bohoffi thanks for all the input and tests on this, very nice!
We could do a kind of simple tab display just along these lines: by using the best guess and ignoring the ambiguity of the many possible positions of the same pitch on a guitar. One could even add a fingering/hand model with some kind of smallest distance metric to get from each note to the next with the shortest distance. This would ask for some nice optimization algorithm :-) 
But for the standard way we would need the following:

- reading the technical string and fret from xml
- add the properties to the Note object
- handing that technical info over to vexflow and adapt the VexFlowMeasure.draw() method to draw notes and tab
- adapt the calculation of the GraphicalMeasure height to include if a tab is added or not 

So for now i would just make the GraphicalMeasure have the full height of the notes-stave and the tab-stave together including the distance in between, so that its BoundingBox will span over both staves. With that the current simple y-spacing calculation should again work. (For the y-spacing calculation we later need the more sophisticated solution which calculates the height profile of all objects reaching out of any staff. We have that implemented but not ported to osmd yet)
@bohoffi would you wanna try that? Would of course be great!

**Comment by bohoffi** [comment]
@matt-uib I really would like to try but I think I will be done not before next week...

**Comment by matt-uib** [comment]
Cool, great! Whenever you can spend your time on it!

**Comment by yipcma** [comment]
I'm interested in fingering display as well. I've some xml from musescore with fingering. It'd be super helpful if anyone could point me to how i may display them. Thanks a lot in advance.

**Comment by sebastianhaas** [comment]
@yipcma I think @bohoffi and @matt-uib  already pointed out a feasible approach, let's see if anyone finds time to put in some more work :)

**Comment by matt-uib** [comment]
I found the time and implemented a first version rendering the Tabs.
Please checkout https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/tree/feature/Tabs
and try with an xml with Tabs for a sneak preview. (There is still some work to do here of course..)

**Comment by megatronofdeath3000** [comment]
Hello people!
Is there any update on this issue?
@bohoffi would you please explain or give some detailed information about where I need to call 
Vex.Flow.Formatter.FormatAndDrawTab(context, tabStave, stave, tabNotes, staveNotes, true, { align_rests: true }); to get tabs like on your first screenshot?
@matt-uib would you please explain how to use your version rendering the tabs? Is that link actual? https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/tree/feature/Tabs

**Comment by mdubey96** [comment]
I am also very interested in rendering Tabs - is there any documentation for the work done by @matt-uib ?

**Comment by matt-uib** [comment]
We currently don't have the resources to work on the tabs, but we would very much appreciate somebody from the community to continue here.
There is no explicit documentation, but we are of course here for questions if somebody would start to work on it.

**Comment by sschmidTU** [comment]
And of course the code is often its best documentation ;)

**Comment by sschmidTU** [comment]
If any work is done on this, the develop branch should be merged into the tabs branch first,
and that is not so easy, because the tabs branch is ancient.
A lot ways to access OSMD variables and methods have changed, and a simple merge will create many conflicts. It's probably easier to merge tabs into develop (on a new branch created from develop) than the other way round, but even then the automatic merge changes could sneak in outdated code.
Maybe the easiest and safest way would be to look at the commits in the tab branch and integrate everything into OSMD by hand, considering the differences in how OSMD has changed.

@megatronofdeath3000 yes, that tabs branch is the current state of the matter.

**Comment by mdubey96** [comment]
I see - does the current develop branch support any version of tab related rendering?
For example, I see a function called "createFingerings" - is this a tab related function?

**Comment by mdubey96** [comment]
Thank you for clarifying.
From the looks of it the tabs branch is only 6 commits ahead but 543 commits behind the current develop.
Is it these 6 commits that need to be merged/fixed or the 543 that need to be done?

**Comment by MilesMorel** [comment]
Hi all - amazing work on OSMD!! 
So I tried to merge and that was awful but cherry picking the tabs commits into develop looks pretty promising. The only file I'm struggling on is VexFlowTabMeasure.ts and how it should implement the new interfaces (e.g. GraphicalNote). I haven't had enough time to fully understand what's going on here but if someone can help me get this over the line I think it's 90% of the way there. I'll def donate if we can get this feature merged asap and I'm willing to work on it if you can help me see what to do. Anyone interested?

**Comment by MilesMorel** [comment]
@matt-uib Alternatively, what would it take to get this done by say January 15th or sooner? I understand your time is valuable. I'm willing to pay to get this feature asap.

**Comment by matt-uib** [comment]
Dear @MilesMorel ,
sorry - i was just knocked out a week due to a heavy flu and am still not fully fine again.
Can you please write to office@phonicscore.com or directly to me: m.uiberacker@phonicscore.com to get in touch for the details you want to achieve?

**Comment by MilesMorel** [comment]
Ah sorry to hear that you got sick. Thanks for the contact 👍Wishing you a fast recovery!

**Comment by matt-uib** [comment]
So.., dear friends of tab rendering!
i created now a new branch from develop where i cherry-picked the changes from the old tab branch and fixed also the aligning with rest notes.

There are some little problems currently:
- key signature is also rendered into the Tab
- Tuplet numbers are also rendered above the Tab with a rather large y-distance

**Comment by HerbertYu** [comment]
It's normal when loading a tablature for the first time. But after re-rendering, vertical lines display at wrong positions.
![image](https://user-images.githubusercontent.com/9133614/71158705-5e822580-227f-11ea-96ae-0885e49d3938.png)

**Comment by HerbertYu** [comment]
Finally I found the solution.

VexFlowMeasure.ts

    public clean(): void {
        this.vfTies.length = 0;
        this.connectors = [];
        // Clean up instructions
        this.resetLayout();
        
        // Fix the bug here
        if(this.tabMeasure != undefined)
            (this.tabMeasure as VexFlowTabMeasure).clean();
    }

**Comment by matt-uib** [comment]
Dear @HerbertYu thanks very much for the finding! Completely right! Didn't check that and overlooked it. Will be added.

**Comment by MilesMorel** [comment]
@matt-uib Thanks so much for all this! Really looking forward to the next update!!

**Comment by MilesMorel** [comment]
@matt-uib it's my understanding that this work is done but it looks like it hasn't been merged - I'm concerned it will get stale again after all your good work. Can you please merge your tabs work into the main branch? 🙏 (or maybe I'm missing something...)

**Comment by matt-uib** [comment]
@MilesMorel yes you are right of course- we will merge it. I'm just once again knocked out by some flu currently...

**Comment by MilesMorel** [comment]
@matt-uib Oh no! Are you ok?? I really hope it isn't covid-19. My co-founder has it we think :( He's been very sick.

**Comment by mdubey96** [comment]
Hey @matt-uib, hope you are feeling better! Is there any progress/updates on the tab branch?

Best,
MD

**Comment by matt-uib** [comment]
Yes, thanks i'm healthy again already (no covid-19) - problem is that my wife is still not and i'm kept busy with our little one and doing all kinds of other stuff...
Hope to be soon back to work...

**Comment by MilesMorel** [comment]
@sschmidTU I paid a hefty sum to get this done by mid-January and now its mid-April and it is still totally unusable. I am sincerely sympathetic to @matt-uib but I made business decisions around this being done and now my livelihood is at stake as my back-list of clients require this feature. Isn't there anyone else at PhonicScore who can merge this? I understand you guys are busy, but you did take my donation with the promise of providing me with this feature months ago and now I'm really in a bind, waiting on it for months. Please help.

**Comment by sschmidTU** [comment]
@MilesMorel Hey, sorry for your troubles, we didn't know it was so urgent for you. We now have the branch basically merged and are working on testing it and creating a new release, so you can use it. It should be published very soon.

**Comment by sschmidTU** [comment]
Does anyone have a few Musicxml Tab samples? They aren't easy to find. There was only one posted in this issue, and it's not correct, it's missing a TAB keyword.

**Comment by MilesMorel** [comment]
Thank you @sschmidTU I look forward to that. MusicXML with tabs can vary a bit depending on how the tablature is added and from which program. Here are a few examples.

[tab examples.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/4476318/tab.examples.zip)

**Comment by sschmidTU** [comment]
Thank you! The A Major scale and Greensleeves look good,
Old_Joe_Clark renders in violin clef instead of tablature, probably because it's also missing the `<clef>` attribute, e.g. in Greensleeves:
```
        <clef number="2">
          <sign>TAB</sign>
          <line>5</line>
          </clef>
```

I suppose we can simply make the system a tab system when we have `<string>` and `<fret>` tags, like other software seems to imply.

**Comment by sschmidTU** [comment]
Wait, Old_Joe_Clark seems to simply be written for Classical Guitar, in Octave Violin clef, no tablature here, no `<string>` and `<fret>` tags either.

**Comment by MilesMorel** [comment]
@sschmidTU My mistake on Old_Joe_Clark - I have a tabbed and tab-free version of that one. I'll see if I can get some more examples if it helps

**Comment by MilesMorel** [comment]
Here are some more simple tunes that use a grand-staff with TAB.
 
[Archive.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/4476522/Archive.zip)

**Comment by MilesMorel** [comment]
@sschmidTU I think the "Study for Guitar 1" example I sent is a pretty full-featured "real-world" example because it has multiple voices and polyphony as well as repeats and a grand-staff.

**Comment by sschmidTU** [comment]
No problem, thanks again! These all render fine.
I just merged the tabs branch into develop in #716 :)
I'm preparing for a new OSMD release now, so if you have any last minute requests, you can test it out on the develop branch, but we're probably releasing another OSMD version soon anyways because of all the recent Vexflow improvements we'll integrate.

**Comment by sschmidTU** [comment]
I'm looking for a sample tabulature MusicXML that we can use for our visual regression tests in the future.
It would basically be public domain of course, under our BSD license.
It would be nice to have something with polyphony (multiple notes simultaneously), and maybe also ties and fingerings.

**Comment by sschmidTU** [comment]
@MilesMorel is it okay if we use the Greensleeves sample as a demo and testing sample for OSMD for now?

**Comment by sschmidTU** [comment]
Sample of the guitar study rendered in OSMD:
![image](https://user-images.githubusercontent.com/33069673/79261883-be38b800-7e90-11ea-8c5e-39f4b3ae938f.png)

**Comment by sschmidTU** [comment]
OSMD 0.7.4 is released, supporting Tabs! Try the [public demo](https://opensheetmusicdisplay.github.io/demo/).
We may release another OSMD version 0.7.4a with a few package updates soon, and then another one with the latest Vexflow version of course.

**Comment by MilesMorel** [comment]
@sschmidTU Excellent work, thank you for getting on this today! That guitar study I shared is in the public domain from IMSLP and Greensleeves is also in the public domain. Feel free to use either in your demo.

**Comment by sschmidTU** [comment]
@MilesMorel do you have a link to this guitar study from IMSLP and maybe a list of other tabs from IMSLP? Just want to check the sources.

**Comment by MilesMorel** [comment]
From IMSLP:
https://imslp.org/wiki/4_Pieces_for_Guitar_(Hayden%2C_Winslow_Lewis)

**Comment by MilesMorel** [comment]
The only thing that I don't have permission for you to publish is that A Major scale pattern I shared. Everything else I shared is in the public domain to my knowledge. We spruced up the arrangements but feel free to share a couple of them.

**Comment by sschmidTU** [comment]
Thank you! I think that extract of the guitar study is perfect.

**Comment by sschmidTU** [comment]
OSMD 0.7.5 is now released, with just a few additions to 0.7.4, but for tabs, we now have a demo and testing sample you can [select from the sample list](https://opensheetmusicdisplay.github.io/demo/), OSMD Function Test - Tabulature:

![image](https://user-images.githubusercontent.com/33069673/79290330-21940b80-7ecc-11ea-8254-45082b80696b.png)

**Comment by MilesMorel** [comment]
Great work! This is working beautifully so far 👍
